File: Sources/Compiler/Lexer/Lexer.swift
//
//  Lexer.swift
//  Blitz3DCompiler
//
//  Lexical analyzer for Blitz3D BASIC
//

public struct Lexer {
    private let source: String
    private var currentIndex: String.Index
    private var currentLine: Int
    private var currentColumn: Int
    private let sourceFile: String
    private let lineMap: [Int: (file: String, line: Int)]?
    private var lineStarts: [String.Index]  // For line lookup
    
    public init(source: String, sourceFile: String = "unknown", lineMap: [Int: (file: String, line: Int)]? = nil) {
        self.source = source
        self.sourceFile = sourceFile
        self.lineMap = lineMap
        self.currentIndex = source.startIndex
        self.currentLine = 1
        self.currentColumn = 1
        self.lineStarts = [source.startIndex]
        
        CompilerLogger.trace("DEBUG_LEXER: Initialized with source length: \(source.count) characters")
        
        // Pre-compute line starts for error reporting
        var idx = source.startIndex
        while idx < source.endIndex {
            if source[idx] == "\n" {
                lineStarts.append(source.index(after: idx))
            }
            idx = source.index(after: idx)
        }
        
        CompilerLogger.trace("DEBUG_LEXER: Found \(lineStarts.count) lines")
    }
    
    public mutating func nextToken() -> Token {
        skipWhitespace()
        
        let startLine = currentLine
        let startColumn = currentColumn
        
        guard currentIndex < source.endIndex else {
            CompilerLogger.trace("DEBUG_LEXER: Reached EOF at line \(currentLine), column \(currentColumn)")
            return makeToken(.endOfFile, text: "", line: currentLine, column: currentColumn)
        }
        
        let ch = source[currentIndex]
        
        if ch == "\n" {
            advance()
            currentLine += 1
            currentColumn = 1
            return makeToken(.newline, text: "\n", line: startLine, column: startColumn)
        }
        
        // Comments
        if ch == "'" || ch == ";" {
            skipComment()
            return nextToken()
        }

        // SCPCB/IDE metadata lines can begin with "=;" (treat as a comment line).
        if ch == "=" && startColumn == 1, let next = peek(offset: 1), next == ";" {
            advance() // consume '='
            skipComment() // now positioned at ';'
            return nextToken()
        }
        
        // String literals
        if ch == "\"" {
            return readStringLiteral(startLine: startLine, startColumn: startColumn)
        }
        
        // Numbers (including floats like .01)
        if ch.isASCII && ch.isNumber {
            return readNumber(startLine: startLine, startColumn: startColumn)
        }
        
        // Floats starting with . (like .01, .5)
        if ch == "." {
            if let next = peek(offset: 1), next.isASCII && next.isNumber {
                return readNumber(startLine: startLine, startColumn: startColumn)
            }
        }
        
        // Hex literals
        if ch == "$" {
            return readHexLiteral(startLine: startLine, startColumn: startColumn)
        }

        // Binary literals (%10101)
        if ch == "%" {
            // Check if followed by binary digits (0 or 1)
            if let next = peek(offset: 1), next == "0" || next == "1" {
                return readBinaryLiteral(startLine: startLine, startColumn: startColumn)
            }
            // Otherwise fall through - % will be handled as type suffix or error
        }

        // Identifiers and keywords
        if ch.isASCII && (ch.isLetter || ch == "_") {
            return readIdentifierOrKeyword(startLine: startLine, startColumn: startColumn)
        }
        
        // Operators and punctuation
        return readOperatorOrPunctuation(startLine: startLine, startColumn: startColumn)
    }
    
    public mutating func tokenize() -> [Token] {
        var tokens: [Token] = []
        
        while true {
            let token = nextToken()
            tokens.append(token)
            
            if token.type == .endOfFile {
                break
            }
        }
        
        return tokens
    }
    
    // MARK: - Private Methods
    
    private mutating func skipWhitespace() {
        while currentIndex < source.endIndex {
            let ch = source[currentIndex]
            
            if ch.isWhitespace && ch != "\n" {
                advance()
            } else {
                break
            }
        }
    }
    
    private mutating func advance() {
        if currentIndex < source.endIndex {
            currentIndex = source.index(after: currentIndex)
            currentColumn += 1
        }
    }
    
    private func peek(offset: Int = 0) -> Character? {
        let idx = source.index(currentIndex, offsetBy: offset, limitedBy: source.endIndex)
        return idx != nil && idx! < source.endIndex ? source[idx!] : nil
    }
    
    private mutating func skipComment() {
        while currentIndex < source.endIndex {
            let ch = source[currentIndex]
            if ch == "\n" {
                break
            }
            advance()
        }
    }
    
    private mutating func readStringLiteral(startLine: Int, startColumn: Int) -> Token {
        advance()  // Skip opening quote
        
        var text = ""
        while currentIndex < source.endIndex {
            let ch = source[currentIndex]
            
            if ch == "\"" {
                // Check for escaped quote ("")
                if let next = peek(offset: 1), next == "\"" {
                    text.append("\"")
                    advance()
                    advance()
                } else {
                    advance()  // Skip closing quote
                    break
                }
            } else if ch == "%" || ch == "#" || ch == "$" {
                // Type suffixes can appear in strings (SCPCB quirk)
                text.append(ch)
                advance()
            } else {
                text.append(ch)
                advance()
            }
        }
        
        return makeToken(.stringLiteral, text: text, line: startLine, column: startColumn)
    }
    
    private mutating func readNumber(startLine: Int, startColumn: Int) -> Token {
        var text = ""
        var hasDecimal = false
        
        // Handle numbers starting with . (like .01)
        if source[currentIndex] == "." {
            hasDecimal = true
            text.append(".")
            advance()
        }
        
        while currentIndex < source.endIndex {
            let ch = source[currentIndex]
            
            if ch.isASCII && ch.isNumber {
                text.append(ch)
                advance()
            } else if ch == "." && !hasDecimal {
                hasDecimal = true
                text.append(ch)
                advance()
            } else if (ch == "e" || ch == "E") && !text.contains("e") && !text.contains("E") {
                text.append(ch)
                advance()
                if let next = peek(), next == "+" || next == "-" {
                    text.append(next)
                    advance()
                }
            } else {
                break
            }
        }
        
        // Check for type suffix
        if let next = peek(), next == "#" || next == "%" {
            text.append(next)
            advance()
        }
        
        if hasDecimal || text.contains("e") || text.contains("E") {
            return makeToken(.floatLiteral, text: text, line: startLine, column: startColumn)
        } else {
            return makeToken(.integerLiteral, text: text, line: startLine, column: startColumn)
        }
    }
    
    private mutating func readHexLiteral(startLine: Int, startColumn: Int) -> Token {
        advance() // Skip $

        var text = ""
        while currentIndex < source.endIndex {
            let ch = source[currentIndex]
            if ch.isASCII && (ch.isNumber || (ch >= "a" && ch <= "f") || (ch >= "A" && ch <= "F")) {
                text.append(ch)
                advance()
            } else {
                break
            }
        }

        // Convert to integer
        if let value = Int(text, radix: 16) {
            return makeToken(.integerLiteral, text: String(value), line: startLine, column: startColumn)
        }

        return makeToken(.error, text: "$" + text, line: startLine, column: startColumn)
    }

    private mutating func readBinaryLiteral(startLine: Int, startColumn: Int) -> Token {
        advance() // Skip %

        var text = ""
        while currentIndex < source.endIndex {
            let ch = source[currentIndex]
            if ch == "0" || ch == "1" {
                text.append(ch)
                advance()
            } else {
                break
            }
        }

        // Convert to integer
        if let value = Int(text, radix: 2) {
            return makeToken(.integerLiteral, text: String(value), line: startLine, column: startColumn)
        }
        
        return makeToken(.error, text: "%" + text, line: startLine, column: startColumn)
    }
    
    private mutating func readIdentifierOrKeyword(startLine: Int, startColumn: Int) -> Token {
        var text = ""
        
        while currentIndex < source.endIndex {
            let ch = source[currentIndex]
            if ch.isASCII && (ch.isLetter || ch.isNumber || ch == "_") {
                text.append(ch)
                advance()
            } else {
                break
            }
        }
        
        // Check for type suffix
        if let next = peek(), next == "%" || next == "#" || next == "$" {
            text.append(next)
            advance()
        }
        
        let lowercase = text.lowercased()
        
        // Look ahead for multi-word keywords
        if lowercase == "end" || lowercase == "else" {
            // Save current position
            let savedIndex = currentIndex
            let savedLine = currentLine
            let savedColumn = currentColumn
            
            // Skip whitespace
            while currentIndex < source.endIndex {
                let ch = source[currentIndex]
                if ch.isWhitespace && ch != "\n" && ch != "\r" {
                    advance()
                } else {
                    break
                }
            }
            
            // Try to read next word
            var nextWord = ""
            while currentIndex < source.endIndex {
                let ch = source[currentIndex]
                if ch.isASCII && (ch.isLetter || ch.isNumber || ch == "_") {
                    nextWord.append(ch)
                    advance()
                } else {
                    break
                }
            }
            
            // Check if the combined phrase is a keyword
            if !nextWord.isEmpty {
                let combined = text + " " + nextWord
                if let keywordType = keywordMap[combined.lowercased()] {
                    CompilerLogger.trace("DEBUG_LEXER: Matched multi-word keyword '\(combined)' at line \(startLine)")
                    return makeToken(keywordType, text: combined, line: startLine, column: startColumn)
                }
            }
            
            // Not a multi-word keyword - rewind to saved position
            CompilerLogger.trace("DEBUG_LEXER: No multi-word match for '\(text)' + '\(nextWord)', rewinding")
            currentIndex = savedIndex
            currentLine = savedLine
            currentColumn = savedColumn
        }
        
        // Check if it's a keyword
        if let keywordType = keywordMap[lowercase] {
            if keywordType == .keywordFunction {
                CompilerLogger.trace("DEBUG_LEXER: Producing .keywordFunction at line \(startLine)")
            }
            return makeToken(keywordType, text: text, line: startLine, column: startColumn)
        }
        
        return makeToken(.identifier, text: text, line: startLine, column: startColumn)
    }
    
    private mutating func readOperatorOrPunctuation(startLine: Int, startColumn: Int) -> Token {
        let ch = source[currentIndex]
        advance()
        
        // Two-character operators - check currentIndex is valid BEFORE accessing source
        if currentIndex < source.endIndex {
            let twoChar = String(ch) + String(source[currentIndex])
            
            switch twoChar {
            case "<>":
                advance()
                return makeToken(.notEquals, text: twoChar, line: startLine, column: startColumn)
            case "<=":
                advance()
                return makeToken(.lessThanOrEqual, text: twoChar, line: startLine, column: startColumn)
            case "=<":
                advance()
                return makeToken(.lessThanOrEqual, text: twoChar, line: startLine, column: startColumn)
            case ">=":
                advance()
                return makeToken(.greaterThanOrEqual, text: twoChar, line: startLine, column: startColumn)
            case "=>":
                advance()
                return makeToken(.greaterThanOrEqual, text: twoChar, line: startLine, column: startColumn)
            default:
                break
            }
        }
        
        // Single character
        switch ch {
        case "+": return makeToken(.plus, text: "+", line: startLine, column: startColumn)
        case "-": return makeToken(.minus, text: "-", line: startLine, column: startColumn)
        case "*": return makeToken(.multiply, text: "*", line: startLine, column: startColumn)
        case "/": return makeToken(.divide, text: "/", line: startLine, column: startColumn)
        case "=": return makeToken(.equals, text: "=", line: startLine, column: startColumn)
        case "<": return makeToken(.lessThan, text: "<", line: startLine, column: startColumn)
        case ">": return makeToken(.greaterThan, text: ">", line: startLine, column: startColumn)
        case "(": return makeToken(.leftParen, text: "(", line: startLine, column: startColumn)
        case ")": return makeToken(.rightParen, text: ")", line: startLine, column: startColumn)
        case "[": return makeToken(.leftBracket, text: "[", line: startLine, column: startColumn)
        case "]": return makeToken(.rightBracket, text: "]", line: startLine, column: startColumn)
        case ",": return makeToken(.comma, text: ",", line: startLine, column: startColumn)
        case "\\": return makeToken(.backslash, text: "\\", line: startLine, column: startColumn)
        case ":": return makeToken(.colon, text: ":", line: startLine, column: startColumn)
        case ".": return makeToken(.period, text: ".", line: startLine, column: startColumn)
        case "^": return makeToken(.power, text: "^", line: startLine, column: startColumn)
        default:
            return makeToken(.error, text: String(ch), line: startLine, column: startColumn)
        }
    }

    private func makeToken(_ type: TokenType, text: String, line: Int, column: Int) -> Token {
        if let map = lineMap, let entry = map[line] {
            return Token(type: type, text: text, line: entry.line, column: column, sourceFile: entry.file)
        }
        return Token(type: type, text: text, line: line, column: column, sourceFile: sourceFile)
    }
    
    // Keyword lookup table (case-insensitive)
    private let keywordMap: [String: TokenType] = [
        "dim": .keywordDim,
        "goto": .keywordGoto,
        "gosub": .keywordGosub,
        "return": .keywordReturn,
        "exit": .keywordExit,
        "if": .keywordIf,
        "then": .keywordThen,
        "else": .keywordElse,
        "endif": .keywordEndIf,
        "end if": .keywordEndIf,
        "elseif": .keywordElseIf,
        "else if": .keywordElseIf,
        "while": .keywordWhile,
        "wend": .keywordWend,
        "for": .keywordFor,
        "to": .keywordTo,
        "step": .keywordStep,
        "next": .keywordNext,
        "function": .keywordFunction,
        "end function": .keywordEndFunction,
        "type": .keywordType,
        "end type": .keywordEndType,
        "each": .keywordEach,
        "local": .keywordLocal,
        "global": .keywordGlobal,
        "field": .keywordField,
        "const": .keywordConst,
        "select": .keywordSelect,
        "case": .keywordCase,
        "default": .keywordDefault,
        "end select": .keywordEndSelect,
        "repeat": .keywordRepeat,
        "until": .keywordUntil,
        "forever": .keywordForever,
        "data": .keywordData,
        "read": .keywordRead,
        "restore": .keywordRestore,
        "abs": .keywordAbs,
        "sgn": .keywordSgn,
        "mod": .keywordMod,
        "pi": .keywordPi,
        "true": .keywordTrue,
        "false": .keywordFalse,
        "int": .keywordInt,
        "float": .keywordFloat,
        "str": .keywordStr,
        "include": .keywordInclude,
        "new": .keywordNew,
        "delete": .keywordDelete,
        "first": .keywordFirst,
        "last": .keywordLast,
        "insert": .keywordInsert,
        "before": .keywordBefore,
        "after": .keywordAfter,
        "null": .keywordNull,
        "object": .keywordObject,
        "handle": .keywordHandle,
        "and": .keywordAnd,
        "or": .keywordOr,
        "xor": .keywordXor,
        "not": .keywordNot,
        "shl": .keywordShl,
        "shr": .keywordShr,
        "sar": .keywordSar
    ]
}

File: web-ide/src/compiler/lexer.ts
/**
 * Blitz3D Lexer - Tokenization
 *
 * Converts source code into tokens for parsing
 */

export enum TokenType {
  // Literals
  INTEGER = "INTEGER",
  FLOAT = "FLOAT",
  STRING = "STRING",

  // Identifiers and keywords
  IDENTIFIER = "IDENTIFIER",

  // Keywords
  IF = "IF",
  THEN = "THEN",
  ELSE = "ELSE",
  ELSEIF = "ELSEIF",
  ENDIF = "ENDIF",
  FOR = "FOR",
  TO = "TO",
  STEP = "STEP",
  NEXT = "NEXT",
  WHILE = "WHILE",
  WEND = "WEND",
  REPEAT = "REPEAT",
  UNTIL = "UNTIL",
  FOREVER = "FOREVER",
  SELECT = "SELECT",
  CASE = "CASE",
  DEFAULT = "DEFAULT",
  END = "END",
  FUNCTION = "FUNCTION",
  RETURN = "RETURN",
  LOCAL = "LOCAL",
  GLOBAL = "GLOBAL",
  CONST = "CONST",
  DIM = "DIM",
  TYPE = "TYPE",
  FIELD = "FIELD",
  NEW = "NEW",
  DELETE = "DELETE",
  FIRST = "FIRST",
  LAST = "LAST",
  BEFORE = "BEFORE",
  AFTER = "AFTER",
  EACH = "EACH",
  DATA = "DATA",
  READ = "READ",
  RESTORE = "RESTORE",
  INCLUDE = "INCLUDE",
  TRUE = "TRUE",
  FALSE = "FALSE",
  NULL = "NULL",
  GOTO = "GOTO",
  GOSUB = "GOSUB",

  // Operators
  PLUS = "PLUS",
  MINUS = "MINUS",
  MULTIPLY = "MULTIPLY",
  DIVIDE = "DIVIDE",
  MOD = "MOD",
  POWER = "POWER",

  // Comparisons
  EQ = "EQ", // =
  NE = "NE", // <>
  LT = "LT", // <
  LE = "LE", // <=
  GT = "GT", // >
  GE = "GE", // >=

  // Logical
  AND = "AND",
  OR = "OR",
  NOT = "NOT",
  XOR = "XOR",

  // Bitwise
  SHL = "SHL",
  SHR = "SHR",
  SAR = "SAR",

  // Punctuation
  LPAREN = "LPAREN", // (
  RPAREN = "RPAREN", // )
  COMMA = "COMMA", // ,
  COLON = "COLON", // :
  DOT = "DOT", // .
  BACKSLASH = "BACKSLASH", // \

  // Special
  NEWLINE = "NEWLINE",
  EOF = "EOF",

  // Type suffixes
  PERCENT = "PERCENT", // % (integer)
  HASH = "HASH", // # (float)
  DOLLAR = "DOLLAR", // $ (string)
}

export interface Token {
  type: TokenType;
  value: string;
  line: number;
  column: number;
}

export interface LexerError {
  message: string;
  line: number;
  column: number;
}

const KEYWORDS: Record<string, TokenType> = {
  "if": TokenType.IF,
  "then": TokenType.THEN,
  "else": TokenType.ELSE,
  "elseif": TokenType.ELSEIF,
  "endif": TokenType.ENDIF,
  "for": TokenType.FOR,
  "to": TokenType.TO,
  "step": TokenType.STEP,
  "next": TokenType.NEXT,
  "while": TokenType.WHILE,
  "wend": TokenType.WEND,
  "repeat": TokenType.REPEAT,
  "until": TokenType.UNTIL,
  "forever": TokenType.FOREVER,
  "select": TokenType.SELECT,
  "case": TokenType.CASE,
  "default": TokenType.DEFAULT,
  "end": TokenType.END,
  "function": TokenType.FUNCTION,
  "return": TokenType.RETURN,
  "local": TokenType.LOCAL,
  "global": TokenType.GLOBAL,
  "const": TokenType.CONST,
  "dim": TokenType.DIM,
  "type": TokenType.TYPE,
  "field": TokenType.FIELD,
  "new": TokenType.NEW,
  "delete": TokenType.DELETE,
  "first": TokenType.FIRST,
  "last": TokenType.LAST,
  "before": TokenType.BEFORE,
  "after": TokenType.AFTER,
  "each": TokenType.EACH,
  "data": TokenType.DATA,
  "read": TokenType.READ,
  "restore": TokenType.RESTORE,
  "include": TokenType.INCLUDE,
  "true": TokenType.TRUE,
  "false": TokenType.FALSE,
  "null": TokenType.NULL,
  "goto": TokenType.GOTO,
  "gosub": TokenType.GOSUB,
  "mod": TokenType.MOD,
  "and": TokenType.AND,
  "or": TokenType.OR,
  "not": TokenType.NOT,
  "xor": TokenType.XOR,
  "shl": TokenType.SHL,
  "shr": TokenType.SHR,
  "sar": TokenType.SAR,
};

export class Lexer {
  private source: string;
  private pos = 0;
  private line = 1;
  private column = 1;
  private errors: LexerError[] = [];
  private readonly MAX_TOKENS = 100000; // Prevent infinite tokenization

  constructor(source: string) {
    this.source = source;
  }

  tokenize(): { tokens: Token[]; errors: LexerError[] } {
    const tokens: Token[] = [];

    try {
      let tokenCount = 0;

      while (!this.isAtEnd() && tokenCount < this.MAX_TOKENS) {
        this.skipWhitespace();

        if (this.isAtEnd()) break;

        try {
          const token = this.nextToken();
          if (token) {
            tokens.push(token);
            tokenCount++;
          }
        } catch (error) {
          // Record error but continue tokenizing
          const errorMsg = error instanceof Error
            ? error.message
            : String(error);
          this.errors.push({
            message: errorMsg,
            line: this.line,
            column: this.column,
          });

          // Skip to next line to recover
          while (!this.isAtEnd() && this.peek() !== "\n") {
            this.advance();
          }
          if (!this.isAtEnd()) this.advance(); // Skip newline
        }
      }

      if (tokenCount >= this.MAX_TOKENS) {
        this.errors.push({
          message:
            "Maximum token count exceeded (possible infinite loop in source)",
          line: this.line,
          column: this.column,
        });
      }

      tokens.push({
        type: TokenType.EOF,
        value: "",
        line: this.line,
        column: this.column,
      });
    } catch (error) {
      // Catastrophic error - record and return what we have
      const errorMsg = error instanceof Error ? error.message : String(error);
      this.errors.push({
        message: `Fatal lexer error: ${errorMsg}`,
        line: this.line,
        column: this.column,
      });
    }

    return { tokens, errors: this.errors };
  }

  private nextToken(): Token | null {
    const start = this.pos;
    const startLine = this.line;
    const startColumn = this.column;

    const ch = this.peek();

    // Comments
    if (ch === ";") {
      this.skipComment();
      return null;
    }

    // Newlines
    if (ch === "\n") {
      this.advance();
      const token = this.makeToken(
        TokenType.NEWLINE,
        "\n",
        startLine,
        startColumn,
      );
      this.line++;
      this.column = 1;
      return token;
    }

    // Strings
    if (ch === '"') {
      return this.scanString(startLine, startColumn);
    }

    // Numbers
    if (this.isDigit(ch) || (ch === "." && this.isDigit(this.peekNext()))) {
      return this.scanNumber(startLine, startColumn);
    }

    // Identifiers and keywords
    if (this.isAlpha(ch)) {
      return this.scanIdentifier(startLine, startColumn);
    }

    // Operators and punctuation
    switch (ch) {
      case "(":
        this.advance();
        return this.makeToken(TokenType.LPAREN, "(", startLine, startColumn);
      case ")":
        this.advance();
        return this.makeToken(TokenType.RPAREN, ")", startLine, startColumn);
      case ",":
        this.advance();
        return this.makeToken(TokenType.COMMA, ",", startLine, startColumn);
      case ":":
        this.advance();
        return this.makeToken(TokenType.COLON, ":", startLine, startColumn);
      case ".":
        this.advance();
        return this.makeToken(TokenType.DOT, ".", startLine, startColumn);
      case "\\":
        this.advance();
        return this.makeToken(
          TokenType.BACKSLASH,
          "\\",
          startLine,
          startColumn,
        );
      case "+":
        this.advance();
        return this.makeToken(TokenType.PLUS, "+", startLine, startColumn);
      case "-":
        this.advance();
        return this.makeToken(TokenType.MINUS, "-", startLine, startColumn);
      case "*":
        this.advance();
        return this.makeToken(TokenType.MULTIPLY, "*", startLine, startColumn);
      case "/":
        this.advance();
        return this.makeToken(TokenType.DIVIDE, "/", startLine, startColumn);
      case "^":
        this.advance();
        return this.makeToken(TokenType.POWER, "^", startLine, startColumn);
      case "%":
        this.advance();
        return this.makeToken(TokenType.PERCENT, "%", startLine, startColumn);
      case "#":
        this.advance();
        return this.makeToken(TokenType.HASH, "#", startLine, startColumn);
      case "$":
        this.advance();
        return this.makeToken(TokenType.DOLLAR, "$", startLine, startColumn);

      case "=":
        this.advance();
        return this.makeToken(TokenType.EQ, "=", startLine, startColumn);

      case "<":
        this.advance();
        if (this.peek() === ">") {
          this.advance();
          return this.makeToken(TokenType.NE, "<>", startLine, startColumn);
        } else if (this.peek() === "=") {
          this.advance();
          return this.makeToken(TokenType.LE, "<=", startLine, startColumn);
        }
        return this.makeToken(TokenType.LT, "<", startLine, startColumn);

      case ">":
        this.advance();
        if (this.peek() === "=") {
          this.advance();
          return this.makeToken(TokenType.GE, ">=", startLine, startColumn);
        }
        return this.makeToken(TokenType.GT, ">", startLine, startColumn);

      default:
        this.errors.push({
          message: `Unexpected character: '${ch}'`,
          line: startLine,
          column: startColumn,
        });
        this.advance();
        return null;
    }
  }

  private scanString(startLine: number, startColumn: number): Token {
    this.advance(); // Skip opening quote
    const start = this.pos;

    while (!this.isAtEnd() && this.peek() !== '"') {
      if (this.peek() === "\n") {
        this.line++;
        this.column = 0;
      }
      this.advance();
    }

    if (this.isAtEnd()) {
      this.errors.push({
        message: "Unterminated string",
        line: startLine,
        column: startColumn,
      });
    } else {
      this.advance(); // Skip closing quote
    }

    const value = this.source.substring(start, this.pos - 1);
    return this.makeToken(TokenType.STRING, value, startLine, startColumn);
  }

  private scanNumber(startLine: number, startColumn: number): Token {
    const start = this.pos;

    while (this.isDigit(this.peek())) {
      this.advance();
    }

    // Float?
    if (this.peek() === "." && this.isDigit(this.peekNext())) {
      this.advance(); // Skip '.'

      while (this.isDigit(this.peek())) {
        this.advance();
      }

      const value = this.source.substring(start, this.pos);
      return this.makeToken(TokenType.FLOAT, value, startLine, startColumn);
    }

    const value = this.source.substring(start, this.pos);
    return this.makeToken(TokenType.INTEGER, value, startLine, startColumn);
  }

  private scanIdentifier(startLine: number, startColumn: number): Token {
    const start = this.pos;

    while (this.isAlphaNumeric(this.peek()) || this.peek() === "_") {
      this.advance();
    }

    // Include type suffix if present (%, #, $)
    if (this.peek() === "%" || this.peek() === "#" || this.peek() === "$") {
      this.advance();
    }

    const value = this.source.substring(start, this.pos);
    const lowerValue = value.toLowerCase();

    // Check for keyword (without type suffix)
    const baseValue = value.replace(/[%#$]$/, "").toLowerCase();
    const type = KEYWORDS[baseValue] || TokenType.IDENTIFIER;

    return this.makeToken(type, value, startLine, startColumn);
  }

  private skipWhitespace(): void {
    while (!this.isAtEnd()) {
      const ch = this.peek();
      if (ch === " " || ch === "\t" || ch === "\r") {
        this.advance();
      } else {
        break;
      }
    }
  }

  private skipComment(): void {
    while (!this.isAtEnd() && this.peek() !== "\n") {
      this.advance();
    }
  }

  private peek(): string {
    if (this.isAtEnd()) return "\0";
    return this.source[this.pos];
  }

  private peekNext(): string {
    if (this.pos + 1 >= this.source.length) return "\0";
    return this.source[this.pos + 1];
  }

  private advance(): void {
    if (!this.isAtEnd()) {
      this.pos++;
      this.column++;
    }
  }

  private isAtEnd(): boolean {
    return this.pos >= this.source.length;
  }

  private isDigit(ch: string): boolean {
    return ch >= "0" && ch <= "9";
  }

  private isAlpha(ch: string): boolean {
    return (ch >= "a" && ch <= "z") || (ch >= "A" && ch <= "Z") || ch === "_";
  }

  private isAlphaNumeric(ch: string): boolean {
    return this.isAlpha(ch) || this.isDigit(ch);
  }

  private makeToken(
    type: TokenType,
    value: string,
    line: number,
    column: number,
  ): Token {
    return { type, value, line, column };
  }
}
